File: Programmers/Lv.2/피로도.swift
import Foundation

func solution(_ k:Int, _ dungeons:[[Int]]) -> Int {
    let dungeonCount = dungeons.count
    var stack = [Int]()
    var using = Set<Int>()
    var counts = [Int]()
    var currentK = k
    
    func backtracking() {
        guard stack.count < dungeonCount else {
            counts.append(dungeonCount)
            return
        }
        
        for i in 0..<dungeonCount {
            guard !using.contains(i) else { continue }
            guard currentK >= dungeons[i][0] else {
                counts.append(using.count)
                continue
            }
            
            stack.append(i)
            using.insert(i)
            
            currentK -= dungeons[i][1]
            
            backtracking()
            
            currentK += dungeons[i][1]
            
            using.remove(i)
            stack.removeLast()
        }
    }
    
    backtracking()
    
    return counts.max()!
}

File: Programmers/Lv.2/피로도.py
answer = 0

def solution(k, dungeons):
    global answer
    visited = [False] * len(dungeons)
    backtracking(k, 0, dungeons, visited)
    return answer

def backtracking(k, count, dungeons, visited):
    global answer
    if count > answer:
        answer = count
    
    for i in range(len(dungeons)):
        if (not visited[i] and
                k >= dungeons[i][0]):
            visited[i] = True
            dfs(k - dungeons[i][1], count + 1, dungeons, visited)
            visited[i] = False
